### PR TITLE
[python] add return_cvbooster flag to cv func and publish _CVBooster (#283,#2105,#1445)

### PR DESCRIPTION
**Pull request by momijiame**
This PR allows users to get trained boosters from cv function directly.
This feature is useful for using ensemble techniques and OOF predictions.

Add the changes that proposed in #283 discussion.

- Add ``return_cvbooster`` flag to cv function
- Rename ``_CVBooster`` to make public

### CODE REVIEW

**Comment by momijiame** [comment]
Hmm... At a glance, I think this change did not trigger breaking CI.

`/__w/1/s/.ci/test.sh: line 103: pytest: command not found`
https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=6577&view=logs&j=02a2c3ba-81f8-54e3-0767-5d5adbb0daa9&t=720ee3fa-96d4-5b47-dbf4-01607b74ade2&l=4772

**Comment by StrikerRUS** [comment]
> Hmm... At a glance, I think this change did not trigger breaking CI.

Random network error. Fixed. Sorry for the inconvenience.

**Comment by StrikerRUS** [comment]
@momijiame Please add new public class to the documentation: https://github.com/microsoft/LightGBM/blob/master/docs/Python-API.rst.

@matsuken92 Can you please help to review this PR?

**Comment by momijiame** [comment]
@StrikerRUS Thank you quick response! I added the CVBooster to the documentation.

File: python-package/lightgbm/engine.py [hunk 1]
```diff
@@ -276,19 +276,26 @@ def train(params, train_set, num_boost_round=100,
     return booster
 
 
-class _CVBooster(object):
-    """Auxiliary data struct to hold all boosters of CV."""
+class CVBooster(object):
+    """CVBooster in LightGBM.
+
+    Auxiliary data struct to hold and redirect all boosters of CV.
```
Comment (StrikerRUS) [inline python-package/lightgbm/engine.py:282]: Add some words about how one can use this class. We should document that this class has the same methods (members?) as `Booster` class. All method calls are actually performed for underlying Boosters and then all returned results are returned in a list.

File: tests/python_package_test/test_engine.py [hunk 2]
```diff
@@ -735,6 +735,47 @@ def test_cv(self):
...
+        params = {
+            'objective': 'binary',
+            'metric': 'binary_logloss',
+            'verbose': -1,
+        }
+        lgb_train = lgb.Dataset(X_train, y_train)
+        # with early stopping
+        cv_res = lgb.cv(params, lgb_train,
+                        num_boost_round=100,
+                        early_stopping_rounds=10,
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:750]: Decrease values to save CI time.

File: tests/python_package_test/test_engine.py [hunk 3]
```diff
@@ -735,6 +735,47 @@ def test_cv(self):
...
+                        verbose_eval=False,
+                        nfold=3,
+                        return_cvbooster=True)
+        self.assertIn('cvbooster', cv_res)
+        cvb = cv_res['cvbooster']
+        self.assertIsInstance(cvb, lgb.engine.CVBooster)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:756]:
```suggestion
        self.assertIsInstance(cvb, lgb.CVBooster)
```
(Will be available after adding this class into `__all__`.)

**Comment by momijiame** [comment]
@StrikerRUS Thank you so much for the feedback! I reflected the comments.

File: python-package/lightgbm/engine.py [hunk 4]
```diff
@@ -276,19 +276,28 @@ def train(params, train_set, num_boost_round=100,
...
```
Comment (StrikerRUS) [inline python-package/lightgbm/engine.py:282]:
```suggestion
    Auxiliary data structure to hold and redirect all boosters of ``cv`` function.
```

File: python-package/lightgbm/engine.py [hunk 5]
```diff
@@ -276,19 +276,28 @@ def train(params, train_set, num_boost_round=100,
...
+    This class has the same methods as Booster class.
+    All method calls are actually performed for underlying Boosters and then all returned results are returned in a list.
+    """
```
Comment (StrikerRUS) [inline python-package/lightgbm/engine.py:285]: Document available class attributes.

```suggestion
    All method calls are actually performed for underlying Boosters and then all returned results are returned in a list.

    Attributes
    ----------
    boosters : list of Booster
        The list of underlying fitted models.
    best_iteration : int
        The best iteration of fitted model.
    """
```

File: tests/python_package_test/test_engine.py [hunk 6]
```diff
@@ -735,6 +735,47 @@ def test_cv(self):
...
+        # with early stopping
+        cv_res = lgb.cv(params, lgb_train,
+                        num_boost_round=25,
+                        early_stopping_rounds=5,
+                        verbose_eval=False,
+                        nfold=3,
+                        return_cvbooster=True)
+        self.assertIn('cvbooster', cv_res)
+        cvb = cv_res['cvbooster']
+        self.assertIsInstance(cvb, lgb.CVBooster)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_engine.py:756]: More checks.

```suggestion
        self.assertIsInstance(cvb, lgb.CVBooster)
        self.assertIsInstance(cvb.boosters, list)
        self.assertEqual(len(cvb.boosters), 3)
        self.assertTrue(all(isinstance(bst, lgb.Booster) for bst in cvb.boosters))
```

**Comment by momijiame** [comment]
Thank you for the great suggestion. I committed.

**Comment by StrikerRUS** [comment]
Gently ping @matsuken92 as you wanted to give your review for the general idea of this PR.

**Comment by momijiame** [comment]
I noticed that this PR also solves #1445, so I updated the title.
Users can extract feature importances from fitted Boosters.

NOTE: It would be more useful to support `CVBooster` and/or `List[Booster]` for `plot_importance()`

**Comment by mirekphd** [comment]
@momijiame : thank you for this much needed feature!
I seem to have found a rather important omission: did we ensure `best_iteration` is returned for each fold from `CVBooster` when using early stopping? 

After running `lgbm_cv_dict = lgbm.cv(params=params_dict, ..., return_cvbooster=True)`, with `params_dict` containing a custom `early_stopping_rounds` value,
`lgbm_cv_dict['cvbooster'].boosters[k].best_iteration` seems to be missing (equal to -1 for all folds), even though `lgbm_cv_dict['cvbooster'].best_iteration` is returned correctly (e.g. 123) (if would be -1 if early stopping rounds were not specified and early stopping were not used). 

For example:
```
Numbers of iterations (current and best) for re-fitted CV models :

Fold 0:
154
-1

Fold 1:
150
-1

Fold 2:
155
-1

Fold 3:
151
-1

Fold 4:
155
-1

CVBooster best iteration:
123
```

Without best iteration it is not possible to obtain predictions for each fold nor feature importances (which rely on the iterations number). In fact I suspect they may be incorrect now, because they are the same regardless of the value their `iteration` parameter is set to:

```
refit_lgbm_cv_ret_dict['cvbooster'].boosters[k].feature_importance()
# or
refit_lgbm_cv_ret_dict['cvbooster'].boosters[k].feature_importance(iteration=fold_best_iter_num)
# or
refit_lgbm_cv_ret_dict['cvbooster'].boosters[k].feature_importance(iteration=fold_cur_iter_num)

# where:
fold_best_iter_num=refit_lgbm_cv_ret_dict['cvbooster'].boosters[k].best_iteration
fold_cur_iter_num=refit_lgbm_cv_ret_dict['cvbooster'].boosters[k].current_iteration()
```

**Comment by momijiame** [comment]
@mirekphd Thank you for your feedback.
I do not know if it is documented or not, but it is expected behavior. From before, cv() function updates Boosters while internally synchronizing all folds. early_stopping is triggered by the average metric of all folds. Therefore, best_iteration is equal in all folds. IMHO, this behavior makes cross validation conservative.

We could consider setting best_iteration of Boosters to all the same value, but that would break backwards compatibility. Even before return_cvbooster option was added, the technique of retrieving CVBooster (_CVBooster) by using Callback was exist.

Also, if we want to train Boosters on each folds, we can simply use train() function. The documentation could be improved on the above, but I do not think the behavior should be changed.

**Comment by mirekphd** [comment]
> Therefore, best_iteration is equal in all folds.

Not in this case:). Please note that we are talking only about **early stopping** here, where `best_iteration` is unknown until each fold model completes. Only then can you compute average best iteration across all folds and return it as CVBooster's "best iteration". 

To see that each fold uses a different number of iterations its enough to see that we still have only one training pass per model (as opposed to two passes), and can be also seen from my example above, where the training stopped at iterations ranging from 150 to 155 rounds, as reported by `.current_iteration()` for each fold.

The information is already available somewhere, we just need to expose it to the user, to replace of those -1 values, just like in case of `.current_iteration()`. 

Unless of course ligthgbm makes two modeling passes and uses average iterations number from the first pass in the final pass (which would be rather inefficient IMO:)

**Comment by guolinke** [comment]
@mirekphd 
I think in both XGBoost and LightGBM, the CV will use the average scores from all folds, and use this for the early stopping. Therefore, best_iteration is the same in all folds.
I think this is more stable since the average_score is computed over all data samples, not just over the current fold.
